Feature/base qcvv framework

cdbf1 opened the pull request:
In preparation for the XEB/IRB benchmarking routines, this PR implements some common functionality that can form the basis of a wide QCVV within superstaq.

cdbf1 reviewed (commented):
Move qcvv folder to supermarq

dowusu-antwi commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+            A copy of the original circuit with the provided gate interleaved.
+        """
+        qubits = circuit.all_qubits()
+        interleaved_circuit = deepcopy(circuit)
```
I think `cirq.Circuit.copy` is `deep`:
```suggestion
        interleaved_circuit = circuit.copy()
```

dowusu-antwi commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+        """
+        qubits = circuit.all_qubits()
+        interleaved_circuit = deepcopy(circuit)
+        for k in range(len(circuit) - int(not include_final), 0, -1):
+            interleaved_circuit.insert(k, gate(*qubits))
```
Might be a bit messier, so up to you:
```suggestion
        interleaved_circuit.batch_insert(
            [(k, gate(*qubits)) for k in range(len(circuit) - int(not include_final), 0, -1)]
        )
```

richrines1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+            # Add measurement of qubit state in computational basis.
+            sample.circuit += cirq.measure(sample.circuit.all_qubits())
+
+    def submit_ss_jobs(
```
nit: i don't think we need the `ss` here - it should be clear enough that that's how we're submitting them
```suggestion
    def submit_jobs(
```
(but if we do want to keep it, imo we should use `_superstaq_` instead of `_ss_`)

richrines1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+        for sample in tqdm(waiting_samples, "Retrieving jobs"):
+            if sample.job is None:
+                continue
+            job = self._service.get_job(sample.job)
```
jobs do some work to prevent unnecessary api calls/etc - it would be good to hold on to the job objects themselves instead of calling `get_job()` every time we need to check them

richrines1 reviewed (commented):
(adding a few comments/questions)

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+        self._raw_data: pd.DataFrame | None = None
+        "The data generated during the experiment"
+
+        self._results: NamedTuple | None = None
```
I did have a go at this originally, but the results from each different type of experiment have completely different names so their didn't seem to be any benefit to subclassing an empty dataclass. I could have use a dict but wanted immutability and strong typing so went with a NamedTuple.

richrines1 replied:
that makes sense, though imo it would be worth it even if the parent result class just contained a couple boilerplate items like "target"/experiment type/etc. We could used `dataclasses.dataclass(frozen=True)` to enforce immutability

imo it would be nice to have a single abstract result type we could type check against. this might be silly though so your call - im fine with it as is if you prefer

cdbf1 replied:
A good idea to add some boiler plate items, which then makes dataclasses the best option. Thanks - I've updated

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+        else:
+            self.sample_circuits_with_simulator(target, shots)
+
+    def _clean_circuits(self) -> None:
```
I guess you could view this as a bit of defensive coding to make sure that all the circuits have 
the right measurements (in case when someone subclasses they forget to add the final measurements and
then it doesn't).

I'm happy to remove if thats not really our style.

richrines1 replied:
no the defensiveness is great :) i mostly just don't love quietly changing things in different places - maybe we could change this to `_validate_circuits()` and just throw an error if the circuits aren't constructed properly? (and then subclasses could also redefine it if they have different/additional requirements..)

cdbf1 replied:
Yes I see what you mean about quietly changing things. I agree just checking the circuits is a better shout

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+            target_options: Optional configuration dictionary passed when submitting the job.
+        """
+        if self._samples is not None:
+            warnings.warn("Existing results will be overwritten.")
```
Very good point!

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+                to normal running.
+            target_options: Optional configuration kwargs passed when submitting the job.
+        """
+        for sample in tqdm(self.samples):
```
I couldn't work out a way to do this where I could also store data about each circuit within a job that contained multiple circuits. This is a feature that is quite important when doing the data analysis for the implemented experiments.

richrines1 replied:
> store data about each circuit within a job

sorry im probably missing something obvious but not sure i understand - could you point me to where this is happening?

cdbf1 replied:
> the only problem is that each sub-job starts fresh, so we lose some efficiency in terms of api calls. it seems like it would be really nice if we added some methods to facilitate functionality into the job itself (ideally such that it automatically populates their job info with whatever we've already retrieved for the parent job to avoid extra work, or maybe even with a reference to the parent job so that some api calls only have to happen once for the batch? this second version might be a bit more messy than it's worth though)

Yeah, I've was thinking about this over lunch today. I might have a play tomorrow and see if I can come up with some kind of "BatchJob" object that facilitates this. I think that would be the most elegant solution and could potentially be really nice to have more widely. Might take a bit of work though (and would certainly need to be a different ticket)

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,462 @@
+        self,
+        num_circuits: int,
+        layers: Iterable[int],
+        target: cirq.SimulatorBase | str | None = None,  # type: ignore [type-arg]
```
Good suggestion, I think the new separation works better

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,524 @@
+        self._results: QCVVResults | None = None
+        """The attribute to store the results in."""
+
+        self._samples: Sequence[Sample] | None = None
```
In order to allow subclassed experiments to subclass the `Sample` class, I think this needs top be a Sequence for contravariance

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,524 @@
+                raise ValueError("QCVV experiment circuits can only contain terminal measurements")
+            if not sorted(sample.circuit[-1].qubits) == sorted(self.qubits):
+                raise ValueError(
+                    "The terminal measurement in QCVV experiment circuits must measure all qubits."
```
The functions that process the returned counts assume that all qubits have been measured

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,524 @@
+            A named tuple of the final results from the experiment.
+        """
+
+    def _state_probs_to_dict(
```
Agreed, no longer needed

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,529 @@
+        of layers to use, return a list of :class:`Sample` objects that need to be sampled during
+        the experiment.
+
+    #. :meth:`process_probabilities`: Take the probability distribution over the
```
Yeah good shout, I have now made all the non-user methods private. Hopefully the description above, together with the demo notebook are sufficient to demonstrate to the user how to use the experiment framework.

richrines1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,512 @@
+# Copyright 2021 The Cirq Developers
```
do we need the cirq license on these?

cdbf1 replied:
I was copying what seemed to be the pattern to have Cirq headers wherever we import cirq. If thats not the case then we can remove

richrines1 replied:
sounds good. we can leave it for now, and sort it out w/ #983

richrines1 reviewed (commented):
looks great, just a couple more questions

not necessarily in this pr, but we'll probably also want to add a third path for experiments on local hardware: in this case we would still use superstaq to compile (but not submit) all the circuits, and then lets the user run the compiled circuits locally and pass in their counts back in to the experiment for analysis

cdbf1 commented:
> not necessarily in this pr, but we'll probably also want to add a third path for experiments on local hardware: in this case we would still use superstaq to compile (but not submit) all the circuits, and then lets the user run the compiled circuits locally and pass in their counts back in to the experiment for analysis

Definitely a good shout. A hacky way to do this at the moment would be for the user to use the `_build_circuits()` method to build the samples and then process them separately. Adding the option to compile them would be cool too, and I guess ideally a slick handle for the users to pass in connection to their own hardware would be the dream - definitely a separate PR though!

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,512 @@
+        return interleaved_circuit
+
+    @abstractmethod
+    def _plot_results(self) -> None:
```
Yes good point, plot results should be public. I like having it includes as an option when analyzing the results too though.

> alternatively, maybe this is silly but wdyt about making .plot() a method of BenchmarkingResult?

Unfortunately the plotting function needs access to the raw data, which is stored in the experiement not the results

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,512 @@
+        Returns:
+            A copy of the original circuit with the provided gate interleaved.
+        """
+        interleaved_circuit = circuit.copy()
```
Yeah very good point. We will need to think more about this when implementing the different experiments too (not all of them use this interleave function

richrines1 commented:
^ sounds good! we'll definitely need something like this for testbed users (e.g. aqt and qscout) who use superstaq to compile but don't have cloud-accessible machines. having a slick interface for this would be really nice

cdbf1 commented:
> ^ sounds good! we'll definitely need something like this for testbed users (e.g. aqt and qscout) who use superstaq to compile but don't have cloud-accessible machines. having a slick interface for this would be really nice

I've implemented this by separating out the `prepare_experiment` function from the `run` function and added a `compile` method which will compile the circuits to the desired target (with additional optimization options). Then I've added an additional `run_with_callable` method which allows a custom callable to be used to evaluate each circuit which could in principal be used to run circuits on some local device

richrines1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,589 @@
+    # Public Methods  #
+    ###################
+    @abstractmethod
+    def analyse_results(self, plot_results: bool = True) -> ResultsT:
```
kinda silly/low priority question: do we want to enforce any consistency re: UK vs US spellings?

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,589 @@
+        """
+        total = sum(counts.values())
+
+        probabilities = {
+            format(idx, f"0{self.num_qubits}b"): 0.0 for idx in range(2**self.num_qubits)
+        }
+
+        for key, count in counts.items():
+            probabilities[key] = count / total
```
Because of they way these probability dictionaries are treated as records in a dataframe, to avoid NaNs I prefer enforcing that all the bitstrings are present in the dictionary.

richrines1 replied:
ok sounds good. in that case i think we also need to fill in the missing values in the simulator output?

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,589 @@
+            self._run_check()
+        for sample in tqdm(self.samples, desc="Running circuits"):
+            probability = circuit_eval_func(sample.circuit, **kwargs)
+            if sorted(probability.keys()) != sorted(
+                f"{i:0{self.num_qubits}b}" for i in range(2**self.num_qubits)
+            ):
+                raise RuntimeError("Returned probabilities are missing bitstrings.")
```
I like this way of checking that the bit strings are correct, and rather than enforcing that all probabilities are included I've manually set any missing bitstrings to have zero probability

cdbf1 commented on supermarq-benchmarks/supermarq/qcvv/base_experiment.py:
```diff
@@ -0,0 +1,589 @@
+    job: css.Job | None = None
+    """The superstaq job corresponding to the sample. Defaults to None if no job is
+    associated with the sample."""
+    raw_circuit: cirq.Circuit = field(init=False)
```
I've gone for a mix of both here. The Sample object now has explicitly as `raw_circuit` (used in the init) and a `compiled_circuit` attribute. By default we want to use the compiled circuit if it exists, so the `circuit` property returns exactly this. 

In the `run_on_device()` method, I've also now added a line that updates the compiled circuit for each sample by retrieving it from the `css.Job`.

richrines1 replied:
cool, makes sense to me

richrines1 reviewed (approved):
awesome, this looks great to me % that simulator output fix (if necessary) :)

cdbf1 commented:
Thanks for all your help @richrines1 - I think we've got something really nice here!
